Upload from URI (admin only) exception

After filling this url http://raw.githubusercontent.com/dagolden/Data-Password-Common/master/t/data/500worst.txt an exception...

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: -2147483648
        at cz.cuni.mff.ufal.DSpaceApi.convertBytesToHumanReadableForm(DSpaceApi.java:509)
        at org.dspace.app.xmlui.aspect.submission.submit.UploadStep.addBody(UploadStep.java:389)
        at org.dspace.app.xmlui.aspect.submission.StepTransformer.addBody(StepTransformer.java:145)
        at org.dspace.app.xmlui.wing.AbstractWingTransformer.startElement(AbstractWingTransformer.java:223)
        ... 271 more
```

--- a/dspace-api/src/main/java/cz/cuni/mff/ufal/DSpaceApi.java
+++ b/dspace-api/src/main/java/cz/cuni/mff/ufal/DSpaceApi.java
@@ -451,7 +451,7 @@ public class DSpaceApi {
 	}
 
     public static String convertBytesToHumanReadableForm(long bytes) {
-    	int exp = (int)(Math.log(bytes) / Math.log(1024));
+    	int exp = bytes == 0 ? 0 : (int)(Math.log(bytes) / Math.log(1024));
     	String units[] = new String[]{"bytes", "kB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"};
     	double val = bytes / Math.pow(1024, exp);
     	if(val == (int)val)

--- a/dspace-api/src/main/java/org/dspace/submit/step/UploadStep.java
+++ b/dspace-api/src/main/java/org/dspace/submit/step/UploadStep.java
@@ -21,6 +21,11 @@ import javax.servlet.ServletException;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.LaxRedirectStrategy;
 import org.apache.log4j.Logger;
 import org.dspace.app.util.SubmissionInfo;
 import org.dspace.app.util.Util;
@@ -746,16 +751,20 @@ public class UploadStep extends AbstractProcessingStep
  
         String filePath = null;        
         InputStream fileInputStream = null;
-        URI file = null;
-        
+
         try {
         	filePath = (String)request.getAttribute("fileLocal");
         	if(filePath.startsWith("/")) {
         		filePath = "file://" + filePath;
-         	}
-        	file = new URI(filePath);
-        	URL url = file.toURL();
-        	fileInputStream = url.openStream();
+                fileInputStream = new URI(filePath).toURL().openStream();
+         	}else {
+                try (CloseableHttpClient client = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()) {
+                    HttpGet httpGet = new HttpGet(filePath);
+                    try (CloseableHttpResponse getResponse = client.execute(httpGet)) {
+                        fileInputStream = getResponse.getEntity().getContent();
+                    }
+                }
+            }
         }catch(IllegalArgumentException e) {
         	return STATUS_NOT_FOUND;
         }catch(MalformedURLException e) {
@@ -806,7 +815,7 @@ public class UploadStep extends AbstractProcessingStep
             b = bundles[0].createBitstream(fileInputStream);
         }
 
-        File f = new File(file.getPath());
+        File f = new File(filePath);
         b.setName(f.getName());
         b.setSource(filePath);
         b.setDescription(fileDescription);

--- a/dspace-api/src/main/java/org/dspace/submit/step/UploadStep.java
+++ b/dspace-api/src/main/java/org/dspace/submit/step/UploadStep.java
@@ -751,6 +751,8 @@ public class UploadStep extends AbstractProcessingStep
  
         String filePath = null;        
         InputStream fileInputStream = null;
+        CloseableHttpClient client = null;
+        CloseableHttpResponse getResponse = null;
 
         try {
         	filePath = (String)request.getAttribute("fileLocal");
@@ -758,12 +760,10 @@ public class UploadStep extends AbstractProcessingStep
         		filePath = "file://" + filePath;
                 fileInputStream = new URI(filePath).toURL().openStream();
          	}else {
-                try (CloseableHttpClient client = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()) {
-                    HttpGet httpGet = new HttpGet(filePath);
-                    try (CloseableHttpResponse getResponse = client.execute(httpGet)) {
-                        fileInputStream = getResponse.getEntity().getContent();
-                    }
-                }
+                client = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build();
+                HttpGet httpGet = new HttpGet(filePath);
+                getResponse = client.execute(httpGet);
+                fileInputStream = getResponse.getEntity().getContent();
             }
         }catch(IllegalArgumentException e) {
         	return STATUS_NOT_FOUND;
@@ -813,6 +813,15 @@ public class UploadStep extends AbstractProcessingStep
         {
             // we have a bundle already, just add bitstream
             b = bundles[0].createBitstream(fileInputStream);
+        }
+        if(fileInputStream != null){
+            fileInputStream.close();
+        }
+        if(getResponse != null){
+            getResponse.close();
+        }
+        if(client != null){
+            client.close();
         }
 
         File f = new File(filePath);

--- a/utilities/project_helpers/scripts/makefile
+++ b/utilities/project_helpers/scripts/makefile
@@ -213,7 +213,7 @@ fresh_install:
 
 # Grant rights to installation
 grant_rights:
-	chown -R $(TOMCAT_USER):$(TOMCAT_GROUP) $(DIR_INSTALLATION)/
+	chown -RL $(TOMCAT_USER):$(TOMCAT_GROUP) $(DIR_INSTALLATION)/
 	chmod -R g+w $(DIR_INSTALLATION)/
 	
 # maven uses prerequisites to check for mvn version

